First Implementation of Questionnaire Component

biojameskim opened the pull request:
## Overview

- Implemented the Onboarding Quiz component from Mid-fis.
- Created a carousel, questionnaire card, and navigation buttons

## Changes Made

- QuestionnaireCarousel.tsx displays the current QuestionCard
- QuestionCard displays a question and options to choose from
- NavigationButtons allow a user to flip through the questions

## Test Coverage

- Tested by interacting with components on local build.

## Screenshots 

<!-- This could include of screenshots of the new feature / proof that the changes work. -->

<img width="497" alt="Screenshot 2023-03-25 at 11 08 55 PM" src="https://user-images.githubusercontent.com/100223126/227753194-b0ddaed9-de57-498d-b8a4-9ac48fb1a14c.png">

<img width="491" alt="Screenshot 2023-03-25 at 11 09 00 PM" src="https://user-images.githubusercontent.com/100223126/227753193-c3b70374-3d67-4426-abc0-9809f77fc767.png">

<img width="501" alt="Screenshot 2023-03-25 at 11 09 04 PM" src="https://user-images.githubusercontent.com/100223126/227753192-1953a503-7231-420f-a3fe-8f818d123b9c.png">

<img width="489" alt="Screenshot 2023-03-25 at 11 09 09 PM" src="https://user-images.githubusercontent.com/100223126/227753191-e3762031-febd-4a40-9181-09583c71590c.png">

</details>

biojameskim committed:
Initialize tailwind and Implement QuestionCard.tsx

biojameskim committed:
Implement Questionnaire Component

- Implemented QuestionnaireCarousel to display and flip through questions
- Implemented QuestionCard to display question and options
- Implemented NavigationButtons to navigate through questions

biojameskim committed:
Update carousel design

DylanTom commented on frontend/src/components/NavigationButtons.tsx:
Looks great! Maybe export `LeftArrow` and `RightArrow` to an Assets folder so we can change the `svg` if needed

DylanTom replied:
Regarding this, I used the webpack `svgr/webpack` to allow us to directly import SVG icons into our source code. @doctakim

sriramanved commented on frontend/src/components/QuestionCard.tsx:
```diff
@@ -0,0 +1,35 @@
+          </button>
+          <button className="btn bg-gray-500 w-full justify-start normal-case border-transparent hover:border-transparent"
+            onClick={() => console.log('clicked option 5')}>
+            {option5}
```
This is not what we want. For example, if there are questions with less than 5 options, your design would break down here! For all the options corresponding to a specific question index, we want to use the spread operator or .map() function to **map** all of the options to a specific button.

sriramanved commented on frontend/src/components/NavigationButtons.tsx:
```diff
@@ -0,0 +1,32 @@
+        onClick={goNext}
+      >
+        Next {RightArrow}
+      </button>
```
Looks good, please write one-liner comments.

DylanTom commented on frontend/src/components/QuestionCard.tsx:
Button can be extracted to another component (in the same file), but allows for code reuse!

sriramanved commented on frontend/src/components/QuestionnaireCarousel.tsx:
```diff
@@ -0,0 +1,81 @@
+    option3: "I just want to talk to someone.",
+    option4: "idk idk idk idk idk idk idk",
+    option5: "I'm not really sure."
+  }
```
You should test your feature with less than 5 options. Vary the number of options to fully test this.

sriramanved commented on frontend/src/components/QuestionnaireCarousel.tsx:
```diff
@@ -0,0 +1,81 @@
+        option3={questions[currentQuestionIndex].option3}
+        option4={questions[currentQuestionIndex].option4}
+        option5={questions[currentQuestionIndex].option5}
+      />
```
So here, you will have to pass in the question counter, but only map the number of options passed in for that specific question to props.

DylanTom commented on frontend/src/components/QuestionnaireCarousel.tsx:
Great use of comments! For now, the dummy questions are fine,but in the future, this should be able to be handled by a JSON so that questions can be changed and adapted.

sriramanved commented on frontend/src/pages/index.tsx:
```diff
@@ -1,123 +1,24 @@
+        option2='Could be better.'
+        option3="I'm not really sure"
+        option4="I'm not doing too well"
+        option5="I need some help"
```
This should not be part of your PR. This should have only been left in here for testing purposes. Please remove this from the index.tsx page when you make changes to this PR.

DylanTom reviewed (changes requested):
This pull request looks great! I'm glad to see you were able to get a working questionnaire working. Just a few tips and comments on abstraction and code reuse for the future.

sriramanved reviewed (changes requested):
Good start, but please be careful about what the MVP for a ticket you are working on actually is. A key consideration is that not all questions will have the same number of options. If you were unsure to consider that in your ticket for this week, you should clarify next time. That is the main point of the ticket, so I will need you to update this component.

biojameskim committed:
Refactor Questionnaire Component

- Abstracted reused code
- Allowed for QuestionCard to display any amount of questions

biojameskim committed:
Add documentation

biojameskim committed:
Add assets

biojameskim committed:
Merge branch 'onboarding-quiz' of https://github.com/cornellh4i/okb-hope into onboarding-quiz

Archit404Error commented on frontend/src/components/QuestionnaireCarousel.tsx:
```diff
@@ -0,0 +1,75 @@
+      'Tell me more about mental health.'
+    ]
+  }
+]
```
Abstract this out into a dummy `.json` file and then import the json file as `questions` so that you can improve code quality

biojameskim replied:
I used the components in the index.tsx file but Ved said to delete it. Should I just create a new dummy file and use them?

Archit404Error commented on frontend/src/components/QuestionnaireCarousel.tsx:
```diff
@@ -0,0 +1,75 @@
+    // If we're on the first question, we want to stay at the first question.
+    if (currentQuestionIndex == 0) setCurrentQuestionIndex(0)
+    else setCurrentQuestionIndex(currentQuestionIndex - 1);
+  }
```
We can rewrite this more clearly using a ternary and docstring as such:

```suggestion
/**
 * Moves to either the previous question or stays at the first question, if we are currently on the first question
 */
  const goBack = () => setCurrentQuestionIndex(currentQuestionIndex == 0 ? 0 : currentQuestionIndex - 1)
```

Archit404Error commented on frontend/src/components/QuestionnaireCarousel.tsx:
```diff
@@ -0,0 +1,75 @@
+    // If we're on the last question, we want to stay at the last question.
+    if (currentQuestionIndex == maxQuestionCount - 1) setCurrentQuestionIndex(maxQuestionCount - 1)
+    else setCurrentQuestionIndex(currentQuestionIndex + 1);
+  }
```
Once again we can use a ternary:

```suggestion
  // Updates the question index to go to the next question.
  const goNext = () => setCurrentQuestionIndex(currentQuestionIndex == maxQuestionCount - 1 ? maxQuestionCount - 1 : currentQuestionIndex + 1)
```

Archit404Error reviewed (changes requested):
Solid work, just some suggestions to clean up your code a bit. **Also, I don't see any of your components being used anywhere in the code.** Please create some route or dummy page where you actually use the components so we can see what they look like when we run the branch locally (without having to create this dummy page ourselves)

biojameskim committed:
Refactor code and extract json file

Archit404Error committed:
Install dependencies

Archit404Error committed:
Resolve merge conflicts

Archit404Error committed:
Remove babelrc

sriramanved committed:
removed postcss config

sriramanved committed:
Merge branch 'onboarding-quiz' of https://github.com/cornellh4i/okb-hope into onboarding-quiz

sriramanved committed:
fixed tailwind config

swinshu commented on frontend/src/pages/index.tsx:
Is this file supposed to be updated in the PR?

swinshu reviewed (commented):
Everything else seems to be covered by the other comments!

sriramanved committed:
Merge branch 'main' into onboarding-quiz

sriramanved committed:
Merge main into branch

sriramanved committed:
created dummy page

sriramanved reviewed (approved):
Good work on this ticket!
